From adcffe29133ff0ab9db6b6a7555b17212207a92c Mon Sep 17 00:00:00 2001
From: Daniele Palmas <dnlplm@gmail.com>
Date: Tue, 28 Sep 2021 14:41:13 +0200
Subject: [PATCH 08/13] bus: mhi: pci_generic: set wake get/put as no-op for
 FN980 v1 hw

FN980 v1 hw does not properly support wake_db register related
operations, so set wake get/put as no-op.

Remove also unsupported devices.

Signed-off-by: Daniele Palmas <dnlplm@gmail.com>
---
 drivers/bus/mhi/pci_generic.c | 32 +++++++++++++++++++++++++++++++-
 1 file changed, 31 insertions(+), 1 deletion(-)

diff --git a/drivers/bus/mhi/pci_generic.c b/drivers/bus/mhi/pci_generic.c
index 76b378ffb038..1b4be2d20353 100644
--- a/drivers/bus/mhi/pci_generic.c
+++ b/drivers/bus/mhi/pci_generic.c
@@ -479,10 +479,40 @@ static const struct mhi_pci_dev_info mhi_telit_fn980_info = {
 	.dma_data_width = 32
 };
 
+static const struct mhi_channel_config mhi_telit_fn980_hw_v1_channels[] = {
+	MHI_CHANNEL_CONFIG_UL(12, "MBIM", 4, 0),
+	MHI_CHANNEL_CONFIG_DL(13, "MBIM", 4, 0),
+	MHI_CHANNEL_CONFIG_UL(14, "QMI", 16, 0),
+	MHI_CHANNEL_CONFIG_DL(15, "QMI", 16, 0),
+	MHI_CHANNEL_CONFIG_UL(20, "IPCR", 16, 0),
+	MHI_CHANNEL_CONFIG_DL_AUTOQUEUE(21, "IPCR", 16, 0),
+	MHI_CHANNEL_CONFIG_HW_UL(100, "IP_HW0", 128, 2),
+	MHI_CHANNEL_CONFIG_HW_DL(101, "IP_HW0", 128, 3),
+};
+
+static struct mhi_controller_config modem_telit_fn980_hw_v1_config = {
+        .max_channels = 128,
+        .timeout_ms = 20000,
+        .num_channels = ARRAY_SIZE(mhi_telit_fn980_hw_v1_channels),
+        .ch_cfg = mhi_telit_fn980_hw_v1_channels,
+        .num_events = ARRAY_SIZE(mhi_telit_fn980_events),
+        .event_cfg = mhi_telit_fn980_events,
+};
+
+static const struct mhi_pci_dev_info mhi_telit_fn980_hw_v1_info = {
+	.name = "telit-fn980",
+	.fw = "qcom/sdx55m/sbl1.mbn",
+	.edl = "qcom/sdx55m/edl.mbn",
+	.config = &modem_telit_fn980_hw_v1_config,
+	.bar_num = MHI_PCI_DEFAULT_BAR_NUM,
+	.dma_data_width = 32,
+	.sideband_wake = true,
+};
+
 static const struct pci_device_id mhi_pci_id_table[] = {
 	/* FN980 firmware release having Telit subvendor-id */
 	{ PCI_DEVICE_SUB(PCI_VENDOR_ID_QCOM, 0x0306, 0x1C5D, 0x2000),
-		.driver_data = (kernel_ulong_t) &mhi_telit_fn980_info },
+		.driver_data = (kernel_ulong_t) &mhi_telit_fn980_hw_v1_info },
 	/* Modifying also Qualcomm default entry for FN980 firmware release without Telit subvendor-id */
 	{ PCI_DEVICE(PCI_VENDOR_ID_QCOM, 0x0306),
 		.driver_data = (kernel_ulong_t) &mhi_telit_fn980_info },
-- 
2.32.0

